src/tools/genpolicy/src/registry.rs

```diff
@@ -252,130 +256,166 @@ async fn get_image_layers(
+        verity_hash = read_verity_from_store(&cache_file, &layer_digest)?;
+        info!("Using cache file");
+        info!("dm-verity root hash: {verity_hash}");
     }
 
-    if !error {
-        match std::fs::read_to_string(&verity_path) {
-            Err(e) => {
-                error = true;
-                error_message = format!("Failed to read {:?}, error {e}", &verity_path);
-            }
-            Ok(v) => {
-                verity_hash = v;
-                info!("dm-verity root hash: {verity_hash}");
+    // create the layer files
+    if !error && verity_hash == "" {
+        if let Err(e) = create_decompressed_layer_file(
+            client,
+            reference,
+            layer_digest,
+            &decompressed_path,
+            &compressed_path,
+        )
+        .await
+        {
+            error_message = format!(
+                "Failed to create verity hash for {layer_digest}, error {e}"
+            );
+            error = true
+        };
+
+        if !error {
+            match get_verity_hash_value(&decompressed_path) {
+                Err(e) => {
+                    error_message = format!("Failed to get verity hash {e}");
+                }
+                Ok(v) => {
+                    verity_hash = v;
+                    if use_cached_files {
+                        add_verity_to_store(&cache_file, &layer_digest, &verity_hash)?;
+                    }
+                    info!("dm-verity root hash: {verity_hash}");
+                }
             }
         }
     }
 
-    if !use_cached_files {
-        let _ = std::fs::remove_dir_all(&base_dir);
-    } else if error {
-        delete_files(&decompressed_path, &compressed_path, &verity_path);
-    }
-
+    temp_dir.close()?;
     if error {
-        panic!("{error_message}");
+        if Path::new(cache_file).exists() {
+            std::fs::remove_file(&cache_file)?;
+        }
+        panic!("{}", &error_message);
     } else {
         Ok(verity_hash)
     }
 }
 
-async fn create_verity_hash_file(
-    use_cached_files: bool,
-    client: &mut Client,
-    reference: &Reference,
+#[derive(Serialize, Deserialize)]
+struct Layer {
+    layer_digest: String,
+    verity_hash: String,
+}
+
+// the store is a json file that matches layer hashes to verity hashes
+fn add_verity_to_store(
+    cache_file: &str,
     layer_digest: &str,
-    base_dir: &Path,
-    decompressed_path: &Path,
-    compressed_path: &Path,
-    verity_path: &Path,
+    verity_hash: &str,
 ) -> Result<()> {
-    if use_cached_files && decompressed_path.exists() {
-        info!("Using cached file {:?}", &decompressed_path);
+    // open the json file in read mode, create it if it doesn't exist
+    let read_file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(cache_file)?;
+
+    // If the file is empty, start with an empty JSON vector
+    let mut data: Vec<Layer> = if read_file.metadata()?.len() == 0 {
```

Comment:
Is there any harm in calling serde_json::from_reader for an empty file?

Reply:
Yes, an empty file will throw:
```
thread 'main' panicked at src/registry.rs:104:18:
called `Result::unwrap()` on an `Err` value: EOF while parsing a value at line 1 column 0
```
So, the check to make sure it's not empty lets the `ImageLayers` go into an empty `Vec`

Reply:
How about this?

let mut data: Vec<Layer> = if let Ok(vec) = serde_json::from_reader(read_file) {
    vec
} else {
    // Delete the malformed file here if it's present
    Vec::new()
};

Reply:
nice, this is cleaner